src/metabase/notification/send.clj

```diff
@@ -333,28 +332,36 @@
                   pool-size
                   (.build
                    (doto (BasicThreadFactory$Builder.)
-                     (.namingPattern "send-notification-thread-pool-%d"))))]
+                     (.namingPattern "send-notification-thread-pool-%d"))))
+        start-worker! (fn []
+                        (.submit ^ThreadPoolExecutor executor
+                                 ^Callable (fn []
+                                             (while (not (Thread/interrupted))
+                                               (try
+                                                 (let [notification (take-notification! queue)]
+                                                   (send-notification-sync! notification))
+                                                 (catch InterruptedException _
+                                                   (log/info "Notification worker interrupted, shutting down")
+                                                   (throw (InterruptedException.)))
+                                                 (catch Throwable e
+                                                   (log/error e "Error in notification worker")))))))]
     (.addShutdownHook
      (Runtime/getRuntime)
      (Thread. ^Runnable (fn []
-                          (.shutdownNow ^ExecutorService executor)
+                          (.shutdownNow ^ThreadPoolExecutor executor)
                           (try
-                            (.awaitTermination ^ExecutorService executor 30 java.util.concurrent.TimeUnit/SECONDS)
+                            (.awaitTermination ^ThreadPoolExecutor executor 30 java.util.concurrent.TimeUnit/SECONDS)
                             (catch InterruptedException _
                               (log/warn "Interrupted while waiting for notification executor to terminate"))))))
     (dotimes [_ pool-size]
-      (.submit ^ExecutorService executor
-               ^Callable (fn []
-                           (while (not (Thread/interrupted))
-                             (try
-                               (let [notification (take-notification! queue)]
-                                 (send-notification-sync! notification))
-                               (catch InterruptedException _
-                                 (log/info "Notification worker interrupted, shutting down")
-                                 (throw (InterruptedException.)))
-                               (catch Exception e
-                                 (log/error e "Error in notification worker")))))))
+      (start-worker!))
     (fn [notification]
+      ;; ensure there is always a worker to handle the notification
+      (dotimes [_ (- pool-size (.getActiveCount ^ThreadPoolExecutor executor))]
+        (log/debugf "Not enough workers, starting a new one %d/%d"
+                    (.getActiveCount ^ThreadPoolExecutor executor)
+                    pool-size)
+        (start-worker!))
```

Comment:
Now the worker loop only terminates on interrupt, is this really necessary?

Reply:
yea, this is just me being annoyed of this bug and want to be super-extra safe :D . I think it's almost always a no-op anyway.